crates/db_schema/src/impls/activity.rs

```diff
@@ -1,139 +1,39 @@
 use crate::{
   newtypes::DbUrl,
-  schema::activity::dsl::{activity, ap_id},
-  source::activity::{Activity, ActivityInsertForm, ActivityUpdateForm},
-  traits::Crud,
+  source::activity::{ReceivedActivity, SentActivity, SentActivityForm},
   utils::{get_conn, DbPool},
 };
 use diesel::{dsl::insert_into, result::Error, ExpressionMethods, QueryDsl};
 use diesel_async::RunQueryDsl;
 
-#[async_trait]
-impl Crud for Activity {
-  type InsertForm = ActivityInsertForm;
-  type UpdateForm = ActivityUpdateForm;
-  type IdType = i32;
-  async fn read(pool: &DbPool, activity_id: i32) -> Result<Self, Error> {
+impl SentActivity {
+  pub async fn create(pool: &DbPool, form: SentActivityForm) -> Result<Self, Error> {
+    use crate::schema::sent_activity::dsl::sent_activity;
     let conn = &mut get_conn(pool).await?;
-    activity.find(activity_id).first::<Self>(conn).await
-  }
-
-  async fn create(pool: &DbPool, new_activity: &Self::InsertForm) -> Result<Self, Error> {
-    let conn = &mut get_conn(pool).await?;
-    insert_into(activity)
-      .values(new_activity)
-      .get_result::<Self>(conn)
-      .await
-  }
-
-  async fn update(
-    pool: &DbPool,
-    activity_id: i32,
-    new_activity: &Self::UpdateForm,
-  ) -> Result<Self, Error> {
-    let conn = &mut get_conn(pool).await?;
-    diesel::update(activity.find(activity_id))
-      .set(new_activity)
+    insert_into(sent_activity)
+      .values(form)
       .get_result::<Self>(conn)
       .await
   }
-  async fn delete(pool: &DbPool, activity_id: i32) -> Result<usize, Error> {
-    let conn = &mut get_conn(pool).await?;
-    diesel::delete(activity.find(activity_id))
-      .execute(conn)
-      .await
-  }
-}
 
-impl Activity {
-  pub async fn read_from_apub_id(pool: &DbPool, object_id: &DbUrl) -> Result<Activity, Error> {
+  pub async fn read_from_apub_id(pool: &DbPool, object_id: &DbUrl) -> Result<Self, Error> {
+    use crate::schema::sent_activity::dsl::{ap_id, sent_activity};
     let conn = &mut get_conn(pool).await?;
-    activity
+    sent_activity
       .filter(ap_id.eq(object_id))
       .first::<Self>(conn)
       .await
   }
 }
 
-#[cfg(test)]
-mod tests {
-  use super::*;
-  use crate::{
-    newtypes::DbUrl,
-    source::{
-      activity::{Activity, ActivityInsertForm},
-      instance::Instance,
-      person::{Person, PersonInsertForm},
-    },
-    utils::build_db_pool_for_tests,
-  };
-  use serde_json::Value;
-  use serial_test::serial;
-  use url::Url;
-
-  #[tokio::test]
-  #[serial]
-  async fn test_crud() {
-    let pool = &build_db_pool_for_tests().await;
-
-    let inserted_instance = Instance::read_or_create(pool, "my_domain.tld".to_string())
+impl ReceivedActivity {
+  pub async fn create(pool: &DbPool, ap_id_: DbUrl) -> Result<Self, Error> {
+    use crate::schema::received_activity::dsl::{ap_id, received_activity};
+    let conn = &mut get_conn(pool).await?;
+    // TODO: use exists() first to avoid spamming log with conflicts
```

Comment:
You can use Insert.[on_conflict_do_nothing()](https://docs.rs/diesel/latest/diesel/query_builder/struct.InsertStatement.html#method.on_conflict_do_nothing). exists() wouldn't be great because it would have a race condition between the check and the insert.